state-commitments/optimized/history_commitment.go

```diff
@@ -24,17 +24,46 @@ func NewCommitment(leaves []common.Hash, virtual uint64) (*Commitment, error) {
 	if len(leaves) == 0 {
 		return nil, errors.New("must commit to at least one leaf")
 	}
+	if virtual == 0 {
+		return nil, errors.New("virtual size cannot be zero")
+	}
 	comm := NewCommitter()
-	root, err := comm.ComputeRoot(leaves, virtual)
+	leavesForProofs := make([]common.Hash, len(leaves))
+	leavesForCommit := make([]common.Hash, len(leaves))
+	copy(leavesForProofs, leaves)
+	firstLeaf := leavesForProofs[0]
+	lastLeaf := leavesForProofs[len(leavesForProofs)-1]
+	for i := 0; i < len(leavesForProofs); i++ {
+		result, err := comm.hash(leavesForProofs[i][:])
+		if err != nil {
+			return nil, err
+		}
+		leavesForCommit[i] = result
+		leavesForProofs[i] = result
+	}
+	var firstLeafProof, lastLeafProof []common.Hash
+	var err error
+	if isPowTwo(virtual) {
+		firstLeafProof, err = comm.computeMerkleProof(0, leavesForProofs, virtual)
+		if err != nil {
+			return nil, err
+		}
+		lastLeafProof, err = comm.computeMerkleProof(virtual-1, leavesForProofs, virtual)
+		if err != nil {
+			return nil, err
+		}
+	}
+	root, err := comm.ComputeRoot(leavesForCommit, virtual)
```

Comment:
There are tree full hashtrees being computed here: for the first leaf, for the last leaf and for the root. All three can be done in a single pass.

Reply:
Im a bit worried about touching the computeroot function's internals to add the different proofs due to the complexity and recursion. These inclusion proofs for first and last leaf are only seldom computed in a challenge, as they are only relevant for powers of two and are bound to a max depth of 26, with most hashes being the precomputed elements.

I like the simplicity of the computeMerkleProof function being a separate thing as it won't be a big performance hit compared to what we had before. ComputeRoot being optimal is the most important IMO